src/ws/cockpitauth.c

```diff
@@ -571,120 +541,239 @@ create_creds_for_authenticated (CockpitAuth *self,
+            {
+              g_debug ("permission denied: %d", (int)code);
+              g_set_error (error, COCKPIT_ERROR, COCKPIT_ERROR_PERMISSION_DENIED,
+                           "Permission denied");
+            }
+          else if (error_str)
+            {
+              g_set_error (error, COCKPIT_ERROR, COCKPIT_ERROR_FAILED,
+                         "Invalid data from %s: %s (%d): %s",
+                         sl->command,
+                         error_str,
+                         (int) code,
+                         error_details);
+            }
+          else
+            {
+              g_set_error (error, G_IO_ERROR, G_IO_ERROR_INVALID_DATA,
+                         "Invalid data from %s: bad result %d",
+                         sl->command,
+                         (int) code);
+            }
         }
     }
-  else if (code == PAM_AUTHINFO_UNAVAIL && g_str_equal (sl->auth_type, "negotiate"))
+
+  build_gssapi_output_header (headers, results);
+
+  if (results)
+    json_object_unref (results);
+  return creds;
+}
+
+static void
+cockpit_auth_local_login_async (CockpitAuth *self,
+                                const gchar *path,
+                                GHashTable *headers,
+                                const gchar *remote_peer,
+                                GAsyncReadyCallback callback,
+                                gpointer user_data)
+{
+  GSimpleAsyncResult *result;
+  LocalLoginData *sl;
+  GBytes *input = NULL;
+  gchar *application;
+  gchar *command = NULL;
+  gchar *type = NULL;
+  gboolean base64_decode = TRUE;
+  const gchar *base64_conf;
+
+  result = g_simple_async_result_new (G_OBJECT (self), callback, user_data,
+                                      cockpit_auth_local_login_async);
+
+  application = cockpit_auth_parse_application (path);
+  type = cockpit_auth_parse_authorization_type (headers);
+
+  if (!type && !gssapi_not_avail)
     {
-      gssapi_not_avail = TRUE;
-      g_debug ("negotiate auth is not available, disabling");
-      g_set_error (error, COCKPIT_ERROR, COCKPIT_ERROR_AUTHENTICATION_FAILED,
-                   "Negotiate authentication not available");
+      type = g_strdup ("negotiate");
+      input = g_bytes_new_static ("", 0);
     }
-  else if (code == PAM_AUTH_ERR || code == PAM_USER_UNKNOWN)
+
+  if (type)
     {
-      g_debug ("authentication failed: %d", (int)code);
-      g_set_error (error, COCKPIT_ERROR, COCKPIT_ERROR_AUTHENTICATION_FAILED,
-                   "Authentication failed");
+      if (cockpit_conf_string (type, "command"))
+        command = g_strdup (cockpit_conf_string (type, "command"));
+      else
+        command = g_strdup (cockpit_ws_session_program);
+
+      if (!input)
+        {
+          base64_conf = cockpit_conf_string (type, "base64-decode");
+          if (base64_conf && (g_ascii_strcasecmp (base64_conf, "false") ||
+              g_ascii_strcasecmp (base64_conf, "0")))
```

Comment:
This needs to be a ```cockpit_conf_boolean()``` function, properly tested. Suggest a different commit for implementing that.